src/roadmap/common.py

```diff
@@ -87,55 +89,133 @@ async def query_rbac(
 
     if not host_permissions:
         raise HTTPException(status_code=403, detail="Not authorized to access host inventory")
 
-    return host_permissions
+    for perm in host_permissions:
+        # Get the list of allowed Group IDs from the attribute filter.
+        for resourceDefinition in perm["resourceDefinitions"]:
+            if not resourceDefinition:
+                # Any record with an empty resourceDefinition means
+                # unrestricted access.
+                # https://insights-rbac.readthedocs.io/en/latest/management.html#resource-definitions
+                return []
+
+            group_list = _get_group_list_from_resource_definition(resourceDefinition)
+            allowed_group_ids.update(group_list)
+
+    return allowed_group_ids
 
 
-# FIXME: This should be cached
 async def query_host_inventory(
     org_id: t.Annotated[str, Depends(decode_header)],
     session: t.Annotated[AsyncSession, Depends(get_db)],
     settings: t.Annotated[Settings, Depends(Settings.create)],
-    permissions: t.Annotated[list[dict], Depends(check_inventory_access)],
+    host_groups: t.Annotated[set[UUID], Depends(get_allowed_host_groups)],
     major: MajorVersion = None,
     minor: MinorVersion = None,
 ):
     if settings.dev:
         org_id = "1234"
 
-    if any(perm.get("resourceDefinitions") for perm in permissions):
-        logger.info(f"RBAC {permissions}")
-        # TODO: Implement workspace filtering
-        raise HTTPException(501, detail="Workspace filtering is not yet implemented")
-
     query = "SELECT * FROM hbi.hosts WHERE org_id = :org_id"
     if major is not None:
         query = f"{query} AND system_profile_facts #>> '{{operating_system,major}}' = :major"
 
     if minor is not None:
         query = f"{query} AND system_profile_facts #>> '{{operating_system,minor}}' = :minor"
 
+    if host_groups:
+        # the hosts database keeps groups data in a JSONB field, with contents
+        # like this:
+        # [
+        #    {
+        #     "account": "123456",
+        #     "created": "2025-01-07T12:58:59.569065+00:00",
+        #     "host_count": 1,
+        #     "id": "f770fbf4-359d-11f0-b21b-5e43c8b8aa2f",
+        #     "name": "GroupTwo",
+        #     "org_id": "1234",
+        #     "ungrouped": false,
+        #     "updated": "2025-01-07T12:59:52.471612+00:00"
+        #    }
+        # ]
+
+        # There is a special case. If None is in host_groups, we must query
+        # for a group that has ungrouped == True.
+        ungrouped_query = "EXISTS (SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj WHERE (group_obj->>'ungrouped')::boolean = true)"
+        # This query searches for any group record with an id that matches any
+        # of our eligible host group ids.
+        grouped_query = "EXISTS (SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj WHERE group_obj->>'id' IN :string_ids)"
```

Comment:
It could be easier to follow if the bind parameter name matched the variable name.
```suggestion
        grouped_query = "EXISTS (SELECT 1 FROM jsonb_array_elements(hosts.groups::jsonb) AS group_obj WHERE group_obj->>'id' IN :host_groups)"
```

Reply:
fixed